src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Diagnostics/RazorDiagnosticsPublisher.cs

```diff
@@ -3,363 +3,272 @@
+                var delegatedParams = new DocumentDiagnosticParams
+                {
+                    TextDocument = new TextDocumentIdentifier { Uri = uriBuilder.Uri },
+                };
+
+                var delegatedResponse = await _clientConnection
+                    .SendRequestAsync<DocumentDiagnosticParams, SumType<FullDocumentDiagnosticReport, UnchangedDocumentDiagnosticReport>?>(
+                        CustomMessageNames.RazorCSharpPullDiagnosticsEndpointName,
+                        delegatedParams,
+                        token)
                     .ConfigureAwait(false);
 
-                if (documentContext is not null)
+                if (delegatedResponse.HasValue &&
+                    delegatedResponse.Value.TryGetFirst(out var fullDiagnostics) &&
+                    fullDiagnostics.Items is not null)
                 {
-                    csharpDiagnostics = await _razorTranslateDiagnosticsService.Value
-                        .TranslateAsync(RazorLanguageKind.CSharp, fullDiagnostics.Items, documentContext, CancellationToken.None)
+                    var documentContext = await _documentContextFactory.Value
+                        .TryCreateAsync(delegatedParams.TextDocument.Uri, projectContext: null, token)
                         .ConfigureAwait(false);
+
+                    if (documentContext is not null)
+                    {
+                        return await _translateDiagnosticsService.Value
+                            .TranslateAsync(RazorLanguageKind.CSharp, fullDiagnostics.Items, documentContext, token)
+                            .ConfigureAwait(false);
+                    }
                 }
             }
-        }
 
-        var razorDiagnostics = result.GetCSharpDocument().Diagnostics;
+            return null;
+        }
+    }
 
-        lock (PublishedRazorDiagnostics)
-        lock (PublishedCSharpDiagnostics)
+    private void StartDelayToClearDocuments()
+    {
+        lock (_documentClosedGate)
         {
-            var filePath = document.FilePath.AssumeNotNull();
-
-            if (PublishedRazorDiagnostics.TryGetValue(filePath, out var previousRazorDiagnostics) && razorDiagnostics.SequenceEqual(previousRazorDiagnostics)
-                && (csharpDiagnostics == null || (PublishedCSharpDiagnostics.TryGetValue(filePath, out var previousCsharpDiagnostics) && csharpDiagnostics.SequenceEqual(previousCsharpDiagnostics))))
+            if (_waitingToClearClosedDocuments)
             {
-                // Diagnostics are the same as last publish
                 return;
             }
 
-            PublishedRazorDiagnostics[filePath] = razorDiagnostics;
-            if (csharpDiagnostics != null)
-            {
-                PublishedCSharpDiagnostics[filePath] = csharpDiagnostics;
-            }
+            _clearClosedDocumentsTask = ClearClosedDocumentsAfterDelayAsync();
+            _waitingToClearClosedDocuments = true;
```

Comment:
Is `_waitingToClearClosedDocuments` really just inverse of `_clearClosedDocumentsTask.IsCompleted`?

In general the `_documentClosedGate` feels strange to me, particularly in the test accessor where it's protecting a field read, but maybe I'm missing something bigger. Seems to me that if the test accessor doesn't want to risk getting a task that is completed, while the next task is about to be set up, isn't the `_publishedDiagnostics` gate enough for that? Seems like one less gate, and one less field would simplify things a little.

But locks and un-awaited async methods are always going to be confusing :)

Reply:
The `_documentClosedGate` and `_publishedDiagnostics` gates protect unrelated data, so felt cleaner to me to keep them separate rather than make them implicitly dependent on each other.

Reply:
But you're right that `_waitingToClearClosedDocuments` is probably unnecessary. I'll take a look at removing that, and then getting rid of `_documentClosedGate`.

Reply:
> The _documentClosedGate and _publishedDiagnostics gates protect unrelated data

Thats fair. It just seemed odd to protect access to a field like that, but maybe I'm getting distracted because it happens to be a `Task`. I don't feel too strongly if you want to leave the gates as they are.